### PR TITLE
Support giga, mega and tera ranges in value formatting

### LINKED ISSUE #620: Scale kWh to Megawatt and Gigawatt if needed
E.g. when displaying yield for a solar charger or PV inverter, using kWh may result in a very big number, so it would be nice to scale it further.

### PR DESCRIPTION
**Pull request by jpetrell**
Fixes #620.

### CODE REVIEW

**Comment by jpetrell** [comment]
Here is the current behavior, please review:

Percentages (precision=0):
NaN (%) -> --%
0 (%) -> 0%
0.4 (%) -> 0.4%
0.55 (%) -> 0.6%
14 (%) -> 14%
15.5 (%) -> 16%
99.3 (%) -> 99.3%
99.7 (%) -> 99.7%
100 (%) -> 100%

Revolutions per minut (precision=0)
NaN (RPM) -> --RPM
0 (RPM) -> 0RPM
0.4 (RPM) -> 0.4RPM
0.55 (RPM) -> 0.6RPM
14 (RPM) -> 14RPM
15.5 (RPM) -> 16RPM
100 (RPM) -> 100RPM
1234 (RPM) -> 1234RPM
12345 (RPM) -> 12kRPM
123456789 (RPM) -> 123MRPM
123456789012 (RPM) -> 123GRPM
123456789012345 (RPM) -> 123TRPM

Ampere-hours (precision=1):
NaN (Ah) -> --Ah
0 (Ah) -> 0.0Ah
0.4 (Ah) -> 0.4Ah
0.54 (Ah) -> 0.5Ah
0.55 (Ah) -> 0.6Ah
14 (Ah) -> 14.0Ah
15.5 (Ah) -> 15.5Ah
100.2 (Ah) -> 100Ah
1234.5 (Ah) -> 1234Ah
12345 (Ah) -> 12.3kAh
123456789 (Ah) -> 124MAh
123456789012 (Ah) -> 124GAh
123456789012345 (Ah) -> 124TAh

Kilowatt-hours (precision=2):
NaN (kWh) -> --kWh
0 (kWh) -> 0.00kWh
0.005 (kWh) -> 0.01kWh
0.05 (kWh) -> 0.05kWh
0.554 (kWh) -> 0.55kWh
0.555 (kWh) -> 0.56kWh
14.123 (kWh) -> 14.1kWh
15.51 (kWh) -> 15.5kWh
100.3134 (kWh) -> 100kWh
1234.5951 (kWh) -> 1235kWh
12345 (kWh) -> 12.3MWh
123456789 (kWh) -> 123GWh
123456789012 (kWh) -> 123TWh

Cubic meters (precision=3):
NaN (m³) -> --m³
0 (m³) -> 0.000m³
0.0005 (m³) -> 0.001m³
0.005 (m³) -> 0.005m³
0.554 (m³) -> 0.554m³
0.5555 (m³) -> 0.556m³
14.1234 (m³) -> 14.12m³
15.5123 (m³) -> 15.51m³
100.3134 (m³) -> 100.3m³
1234.59551 (m³) -> 1235m³
12345.5 (m³) -> 12.35km³
123456789 (m³) -> 123.5Mm³
123456789012 (m³) -> 123.5Gm³
123456789012345 (m³) -> 123.5Tm³

File: CMakeLists.txt [hunk 1]
```diff
@@ -631,13 +631,13 @@ list(APPEND VenusQMLModule_CPP_SOURCES
...
-
-set(Units_CPP_SOURCES
+    src/quantityinfo.h
+    src/quantityinfo.cpp
     src/units.h
     src/units.cpp
 )
-list(APPEND VenusQMLModule_CPP_SOURCES ${Units_CPP_SOURCES})
+
+list(APPEND VenusQMLModule_CPP_SOURCES  ${Units_CPP_SOURCES})
```
Comment (blammit) [inline CMakeLists.txt:640]: Unnecessary change?

File: tests/units/CMakeLists.txt [hunk 2]
```diff
@@ -0,0 +1,46 @@
+#
+# Copyright (C) 2024 Victron Energy B.V.
+# See LICENSE.txt for license information.
+#
+
+cmake_minimum_required(VERSION 3.16)
+project(tst_units LANGUAGES CXX)
```
Comment (blammit) [inline tests/units/CMakeLists.txt:7]: Are you currently building the tests project separately? Should it be built by default as part of the main project?

**Review by blammit — COMMENTED** [review]
Very nice work! I can see this fixes a bunch of rounding/precision issues we had in displaying various values as well. Good to see the tests added for this.

I'm seeing this warning when values are updated: `qrc:/qt/qml/Victron/VenusOS/components/QuantityLabel.qml:23:5: QML QuantityInfo: Binding loop detected for property "value"``

File: src/units.cpp [hunk 3]
```diff
@@ -76,165 +76,264 @@ int Units::defaultUnitPrecision(Victron::VenusOS::Enums::Units_Type unit) const
...
+	case VenusOS::Enums::Units_Temperature_Kelvin:
+	case VenusOS::Enums::Units_Hectopascal:
 	default:
-		return QString();
+		return false;
 	}
 }
 
-Victron::Units::quantityInfo Units::scaledQuantity(
+// Returns the physical quantity as a tuple of strings: { number, unit }.
```
Comment (blammit) [inline src/units.cpp:171]: Can remove this outdated comment

File: CMakeLists.txt (hunk 1 above)
Comment (jpetrell) [reply CMakeLists.txt:640]: Fixed.

File: src/units.cpp (hunk 3 above)
Comment (jpetrell) [reply src/units.cpp:171]: Updated the comment.

**Comment by jpetrell** [comment]
> I'm seeing this warning when values are updated:
> qrc:/qt/qml/Victron/VenusOS/components/QuantityLabel.qml:23:5: QML QuantityInfo: Binding loop detected for property "value"`

Not sure how to avoid that. Hysteresis (the past values affecting the current state) is by definition a binding loop.

File: tests/units/CMakeLists.txt (hunk 2 above)
Comment (jpetrell) [reply tests/units/CMakeLists.txt:7]: I would not build by default with the main project, for most build cases that is just unwanted overhead. I have mostly only ran the tests locally on desktop as there is nothing environment-specific in the formatting functions currently, but you can copy the test to the device and run them there (test pass there also, just tried on Ekrano GX).

Comment (blammit) [reply tests/units/CMakeLists.txt:7]: Is there an easy way to build the tests along with the main project when required (e.g. with some optional flag) so that the tests don't get out-of-date?

Comment (blammit) [reply tests/units/CMakeLists.txt:7]: Yeah, maybe that is too much overhead to include it by default. Is there an easy way to build the tests along with the main project when required (e.g. with some optional flag) so that the tests don't get out-of-date?

**Comment by blammit** [comment]
> > I'm seeing this warning when values are updated:
> > qrc:/qt/qml/Victron/VenusOS/components/QuantityLabel.qml:23:5: QML QuantityInfo: Binding loop detected for property "value"`
> 
> Not sure how to avoid that. Hysteresis (the past values affecting the current state) is by definition a binding loop.

What if we push the update to the next event loop? E.g. maybe using a QueuedConnection for the `connect(this, &QuantityInfo::inputChanged, this, &QuantityInfo::update);` connection would be sufficient?

**Comment by jpetrell** [comment]
> What if we push the update to the next event loop? E.g. maybe using a QueuedConnection

Good idea. Unfortunately that didn't remove the binding loop warning, but when I separated the change signal of value property to own valueChanged from inputChanged the warning went away. Fixed now. Thanks for pushing for quality. :smile:

**Review by blammit — APPROVED** [review]
Looks great! Let's merge :)
